### ISSUE TITLE
cant deploy

### ISSUE THREAD

**Issue by CLAR1US** [josephines1/o-present#1]
why i cant absent in ? its always say "lokasi anda tidak terdeteksi" but my browser not try to ask permission about location. how to make my browser make a permission about location so the app detect my location?

**Comment by josephines1** [comment]
Hi CLAR1US, you can try changing the location access permissions in your browser by going to Settings > Privacy and security > Site settings > Permissions > Location. Make sure the URL of your deployed website is listed under "Allowed to see your location." This method works for Chrome browser.

**Comment by CLAR1US** [comment]
Hi josephines1! first, thank you for making this awesome project!

i just running this project in my local (xampp) and works perfectly! but when i deploy to my vm i found so many stress situation. can you help me about this?

![image](https://github.com/josephines1/o-present/assets/107949634/bec83bf9-8c6e-4314-b88b-7410aa492239)

**Comment by josephines1** [comment]
You might try changing the folder name from "validation" to "Validation". You can find the myCustomValidation class in App\Libraries\validation\myCustomValidation.php. Maybe the error occurs because of sensitivity to capital letters in folder names.

Before change:
App\Libraries\**validation**\myCustomValidation.php

After change:
App\Libraries\**Validation**\myCustomValidation.php

Hope this helps :)

**Comment by CLAR1US** [comment]
Hai Josephines1 ! 

Thankyou for your help, its running !!! But there is a few trouble like :

1. I cant open excel file when i Export it from Daily Absent, it say need repair.
2. It is possible to create employee account without need email activation? if possible, how to make it?

btw once again thank you for making this awesome project! i need your paypal id to give a small donation, or if you indonesian just send me ur bank account number on my linkedin, i was send you a network request

**Comment by josephines1** [comment]
Hello CLAR1US,

Regarding the first issue with the Excel file, you can click "Yes" in the dialog that appears, as shown in the screenshot below. Typically, the repair needed dialog only appears when the website is running on Localhost, but once it has been deployed to a hosting server, this dialog should not appear.

![Screenshot 2024-06-07 175724](https://github.com/josephines1/o-present/assets/147328469/879aa6bb-1885-4796-b55a-4b4052540a81)


For the second issue concerning account activation, creating an employee account without activation is not possible. This is related to the password reset and email change mechanisms, which require the system to send a confirmation to a valid email (the email used to activate the account).

However, if you still want to add an employee account without requiring activation, you can manually add the employee data in the phpMyAdmin database (http://localhost/phpmyadmin). Here’s how:
- Open the 'pegawai' table and insert new data.
- Open the 'users' table and insert new data. When inserting new data in this table, select the 'id_pegawai' for the employee you want to create an account for, and fill in the email, username, and other required fields. For the password, ensure it is hashed using the Myth/Auth CodeIgniter method. Then, set the "active" column to 1.

This is the hash method for passwords used by Myth/Auth CodeIgniter:
```php
password_hash(base64_encode(hash('sha384', "YOUR_PASSWORD", true)), PASSWORD_DEFAULT);
```
Replace "YOUR_PASSWORD" with the desired password string.

Additionally, regarding the network request on LinkedIn, could you please send me a message first? I have accepted all network requests on my LinkedIn. This is just to ensure I can identify your account. Thanks in advance!

**Comment by CLAR1US** [comment]
Hi,

I still can't open excel, even i choose "open and repair"
I was trying using LibreOffice Ubuntu and Microsoft Office LTSC Proffesional PLus 2021 Version 2308 (Build 16731.20316)

![Screenshot 2024-06-08 203805](https://github.com/josephines1/o-present/assets/107949634/cabf6fe7-8218-4c1b-bfc2-63509365d44f)
![Screenshot 2024-06-08 205203](https://github.com/josephines1/o-present/assets/107949634/31e5b084-c889-439d-8050-a3339b066424)
![Screenshot 2024-06-08 205312](https://github.com/josephines1/o-present/assets/107949634/9473ea03-3f68-4572-9670-f33fd3e5e4d9)

Can you help me to fix this? 
Btw what version of Microsoft Office are you using?

**Comment by josephines1** [comment]
Hi, sorry about the bug.

I just pushed an update to the repository to fix it. You can pull the latest from GitHub (with 'git pull') or re-download the files. If you want to update manually, check out the changes I made in the history below.

[https://github.com/josephines1/o-present/commit/8e4b308755af69e5515fb522cc37f9181624378c](https://github.com/josephines1/o-present/commit/8e4b308755af69e5515fb522cc37f9181624378c)

Basically, I added 'exit();' after each line with '$writer->save('php://output');'. I've tried it, and it works. (in Microsoft Excel 2021).

Hope this helps! Thanks for understanding earlier.

**Comment by CLAR1US** [comment]
OMG it works!

But i now i have problem with Mail System, you said :
![image](https://github.com/josephines1/o-present/assets/107949634/fd422a68-81e9-4c59-b9c4-b0944b4dfb2a)

what am I doing right? i don't get it about "code you got from step 2 of the two-step verification" , should i turn on my 2nd authentication first? or just fill with my password?

![image](https://github.com/josephines1/o-present/assets/107949634/8ccdbca4-ac5c-4402-8bb8-0b3e301ce741)

![image](https://github.com/josephines1/o-present/assets/107949634/ed343198-0531-43bc-aa59-d13d1ffb14ac)

**Comment by josephines1** [comment]
Yes, you need to enable 2-step verification for the Google account you'll use to send emails. The code I'm referring to is the app password generated through Google's 2-step verification. To get it, you can follow Google's instructions at the link below.

https://support.google.com/accounts/answer/185833?hl=en

Hope this helps! 😊

**Comment by CLAR1US** [comment]
I found another bug,

1. It show when i want to "reset password" from profile page
![Screenshot 2024-06-09 132158](https://github.com/josephines1/o-present/assets/107949634/e8388c62-2ea0-4032-96c9-99d294e7a692)


2. Head role can't approved/reject "sakit" and "izin" permission, only can "cuti" permission
![Screenshot 2024-06-09 133746](https://github.com/josephines1/o-present/assets/107949634/854ee4c9-892a-408c-95d5-6933dfe8f5ca)


3. When i want "forgot your password" (before login) then i fill token where i get from email, then when i submit it, it show like this
![Screenshot 2024-06-09 135152](https://github.com/josephines1/o-present/assets/107949634/4515740e-3757-4b5d-85a8-7178c18cbfd8)

**Comment by josephines1** [comment]
Hi, sorry about the bug. For issues 1 and 3, I forgot to add these instructions in the Readme. Here are the steps to fix them:

1. Open the file AuthController.php located in vendor\myth\auth\src\Controllers\AuthController.php

2. Change the resendActivateAccount function to this:
```
public function resendActivateAccount($login = false)
    {
        if ($this->config->requireActivation === null) {
            return redirect()->route('login');
        }

        $throttler = service('throttler');

        if ($login == false) {
            if ($throttler->check(md5($this->request->getIPAddress()), 2, MINUTE) === false) {
                return service('response')->setStatusCode(429)->setBody(lang('Auth.tooManyRequests', [$throttler->getTokentime()]));
            }
            $login = urldecode($this->request->getGet('login'));
        }
        $type  = filter_var($login, FILTER_VALIDATE_EMAIL) ? 'email' : 'username';

        $users = model(UserModel::class);

        $user = $users->where($type, $login)
            ->where('active', 0)
            ->first();

        if (null === $user) {
            return redirect()->route('login')->with('error', lang('Auth.activationNoUser'));
        }

        $activator = service('activator');
        $sent      = $activator->send($user);

        if (!$sent) {
            return redirect()->back()->withInput()->with('error', $activator->error() ?? lang('Auth.unknownError'));
        }

        // Success!
        // return redirect()->route('login')->with('message', lang('Auth.activationSuccess'));
        return redirect()->to('/data-pegawai')->with('berhasil', 'Email aktivasi berhasil terkirim');
    }
```

3. Then, also change the attemptForgot function to this:
```
public function attemptForgot($email = false)
    {
        if ($this->config->activeResetter === null) {
            return redirect()->route('login')->with('error', lang('Auth.forgotDisabled'));
        }

        if (!$email) {
            $rules = [
                'email' => [
                    'label' => lang('Auth.emailAddress'),
                    'rules' => 'required|valid_email',
                ],
            ];

            if (!$this->validate($rules)) {
                dd($this->validator->getErrors());
                return redirect()->back()->withInput()->with('errors', $this->validator->getErrors());
            }
        }

        $users = model(UserModel::class);

        if ($email) {
            $user = $users->where('email', $email)->first();
        } else {
            $user = $users->where('email', $this->request->getPost('email'))->first();
        }

        if (null === $user) {
            return redirect()->back()->with('error', lang('Auth.forgotNoUser'));
        }

        // Save the reset hash /
        $user->generateResetHash();
        $users->save($user);

        $resetter = service('resetter');
        $sent     = $resetter->send($user);

        if (!$sent) {
            return redirect()->back()->withInput()->with('error', $resetter->error() ?? lang('Auth.unknownError'));
        }

        return redirect()->route('reset-password')->with('message', lang('Auth.forgotEmailSent'));
    }
```

Now, for issue 2, the Head role can't change the 'ketidakhadiran' status because the requested date is today (June 9, 2024). I set the permissions to only allow updating the absence status for dates after today. So, any absences submitted for today or previous dates can't be updated.

But if you want to allow changes for absences starting today, you can modify it like this:

1. Open the kelola-pengajuan.php file located in app\Views\ketidakhadiran\kelola-pengajuan.php. At line 153, you'll find this code:
```
<?php if ($data->tanggal_mulai <= date('Y-m-d')) : ?>
```

Change it to:

```
<?php if ($data->tanggal_mulai < date('Y-m-d')) : ?>
```

2. Next, open the Ketidakhadiran.php file located in app\Controllers\Ketidakhadiran.php. At line 695, you'll find this code:

```
if ($data_ketidakhadiran->tanggal_mulai <= date('Y-m-d')) {
```

Change it to:

```
if ($data_ketidakhadiran->tanggal_mulai < date('Y-m-d')) {
```

This way, you can update the status for absences starting today.

Thanks for your feedback! Let me know if you have any other issues with the app.
